pic16/hoermann.c

```diff
@@ -178,141 +156,124 @@ static void start_listening(void)
 }
 
-
-static void stop_listening(void)
-{
+static void stop_listening(void) {
   /* Deactivate RS485 and UART receiver */
   NOT_READ_ENABLE = 1;
   RC1STAbits.CREN = 0;
-  
+
   /* Disable receive interrupt */
   RC1IE = 0;
 }
 
-
-void hoermann_init(void)
-{
+void hoermann_init(void) {
   /* UART1 - RS485 */
-  
+
   /* Configure baudrate */
   SP1BRG = RS485_BRGVAL;
-  
+
   BAUD1CONbits.BRG16 = 1;
   TX1STAbits.BRGH = 1;
   /* All other bits in BAUD1CON, RC1STA and TX1STA need not to be changed.
    * Their values at POR/BOR match the required configuration. */
 
   /* Enable UART module */
   RC1STAbits.SPEN = 1;
-  
+
   start_listening();
 }
 
-
-void hoermann_run(void)
-{
+void hoermann_run(void) {
   static uint8_t delay_counter = 0;
-  
-  if(rx_message_ready)
-  {
+
+  if (rx_message_ready) {
     parse_message();
     rx_message_ready = false;
     delay_counter = 3;
     /* Wait 3ms before answering. If not the Supramatic doesn't accept our answer. */
   }
-  if((tx_message_ready)&&(delay_counter == 0))
-  {
+  if ((tx_message_ready) && (delay_counter == 0)) {
     stop_listening();
     start_sending();
     tx_message_ready = false;
   }
-  
-  if((tx_counter == tx_length)&&(TX1STAbits.TRMT == 1)&&(TX1STAbits.TXEN == 1))
-  {
+
+  if ((tx_counter == tx_length) && (TX1STAbits.TRMT == 1) && (TX1STAbits.TXEN == 1)) {
     stop_sending();
     start_listening();
   }
-  
-  if(delay_counter > 0)
-  {
+
+  if (delay_counter > 0) {
     delay_counter--;
   }
 }
 
-
-uint16_t hoermann_get_broadcast(void)
-{
+uint16_t hoermann_get_broadcast(void) {
   return broadcast_status;
 }
 
-
-void hoermann_trigger_action(hoermann_action_t action)
-{
-  switch(action)
-  {
-    case hoermann_action_stop:
-    {
+void hoermann_trigger_action(hoermann_action_t action) {
+  switch (action) {
+    case hoermann_action_stop: {
       slave_respone_data = RESPONSE_STOP;
       break;
     }
-    case hoermann_action_open:
-    {
+    case hoermann_action_open: {
       slave_respone_data = RESPONSE_OPEN;
       break;
     }
-    case hoermann_action_close:
-    {
+    case hoermann_action_close: {
       slave_respone_data = RESPONSE_CLOSE;
       break;
     }
-    case hoermann_action_venting:
-    {
+    case hoermann_action_venting: {
       slave_respone_data = RESPONSE_VENTING;
       break;
     }
-    case hoermann_action_toggle_light:
-    {
+    case hoermann_action_toggle_light: {
       slave_respone_data = RESPONSE_TOGGLE_LIGHT;
       break;
     }
+    case hoermann_action_impulse: {
+      slave_respone_data = RESPONSE_IMPULSE;
+      break;
+    }
+    case hoermann_action_soft_stop: {
+      if ((broadcast_status & 0x60) == 0x40 || (broadcast_status & 0x60) == 0x60) {
+        slave_respone_data = RESPONSE_IMPULSE;
+      }
+      break;
+    }
+    case hoermann_action_soft_stop2: {
+      if ((broadcast_status & 0x60) == 0x40 || (broadcast_status & 0x60) == 0x60) {
+        slave_respone_data = RESPONSE_SOFT_STOP;
+      }
```

Comment:
you are right I kept the stop 0x0000 for backward compatibility. The original project has implemented it like this so I was thinking it's better to keep it unchanged and added the new method at the end.
As updating the Pic16 is not as easy I added both soft stop method so It can then easily decide in esphome code which command to execute.